File: C/CaveEscape/src/cave_escape_10.cpp
#include <stdio.h>
#include "SwinGame.h"

#define GRAVITY 0.08
#define MAX_SPEED 5
#define JUMP_RECOVERY_BOOST 2
#define FOREGROUND_FOREROOF_POLE_SCROLL_SPEED -2
#define BACKGROUND_SCROLL_SPEED -1
#define NUM_POLES 4

typedef struct pole_data
{
  bool score_limiter;
  sprite up_pole, down_pole;
} pole_data;

typedef struct pole_data poles[NUM_POLES];

enum player_state{MENU, PLAY};

typedef struct background_data
{
  sprite foreroof, foreground, background;
} background_data;

typedef struct player
{
  sprite sprite;
  int score;
  bool is_dead;
  player_state state;
} player;

typedef struct game_data
{
  player player;
  background_data scene;
  poles poles;
} game_data;

player get_new_player()
{
  player result;
  result.sprite = create_sprite(bitmap_named("Player"), animation_script_named("PlayerAnimations"));
  sprite_set_x(result.sprite, screen_width() / 2 - sprite_width(result.sprite));
  sprite_set_y(result.sprite, screen_height() / 2);
  sprite_start_animation(result.sprite, "Fly");
  result.score = 0;
  result.is_dead = false;
  result.state = MENU;

  return result;
}

pole_data get_random_poles()
{
  pole_data result;
  result.up_pole = create_sprite(bitmap_named("UpPole"));
  result.down_pole = create_sprite(bitmap_named("DownPole"));
  sprite_set_x(result.up_pole, screen_width() + rnd(1200));
  sprite_set_y(result.up_pole, screen_height() - sprite_height(result.up_pole) -  rnd(bitmap_height(bitmap_named("Foreground"))));
  sprite_set_x(result.down_pole, sprite_x(result.up_pole));
  sprite_set_y(result.down_pole, rnd(bitmap_height(bitmap_named("Foreroof"))));
  sprite_set_dx(result.up_pole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  sprite_set_dx(result.down_pole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  result.score_limiter = true;

  return result;
}

background_data get_new_background()
{
  background_data result;
  result.background = create_sprite(bitmap_named("Background"));
  sprite_set_x(result.background, 0);
  sprite_set_y(result.background, 0);
  sprite_set_dx(result.background, BACKGROUND_SCROLL_SPEED);

  result.foreground = create_sprite(bitmap_named("Foreground"), animation_script_named("ForegroundAminations"));
  sprite_set_x(result.foreground, 0);
  sprite_set_y(result.foreground, screen_height() - sprite_height(result.foreground));
  sprite_set_dx(result.foreground, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  sprite_start_animation(result.foreground, "Fire");

  result.foreroof = create_sprite(bitmap_named("Foreroof"));
  sprite_set_x(result.foreroof, 0);
  sprite_set_y(result.foreroof, 0);
  sprite_set_dx(result.foreroof, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);

  return result;
}

void handle_input(sprite player, player_state* state)
{
  if (key_typed(SPACE_KEY) && (*state = PLAY))
  {
    sprite_set_dy(player, sprite_dy(player) - JUMP_RECOVERY_BOOST);
  }
  else if (key_typed(SPACE_KEY))
  {
    *state = PLAY;
  }
}

void check_for_collisions(game_data *game)
{
  int i;

  if (sprite_collision(game->player.sprite, game->scene.foreground) || sprite_collision(game->player.sprite, game->scene.foreroof))
  {
    game->player.is_dead = true;
    return;
  }

  for (i = 0; i < NUM_POLES; i++)
  {
    if (sprite_collision(game->player.sprite, game->poles[i].up_pole) || sprite_collision(game->player.sprite, game->poles[i].down_pole))
    {
      game->player.is_dead = true;
      return;
    }
  }
}

void reset_pole_data(pole_data *poles)
{
  free_sprite(poles->up_pole);
  free_sprite(poles->down_pole);
  *poles = get_random_poles();
}

void reset_player(player *player)
{
  free_sprite(player->sprite);
  *player = get_new_player();
}

void reset_game(game_data *game)
{
  int i;

  reset_player(&game->player);
  for (i = 0; i < NUM_POLES; i++)
  {
    reset_pole_data(&game->poles[i]);
  }
}

void update_velocity(sprite player)
{
  sprite_set_dy(player, sprite_dy(player) + GRAVITY);

  if (sprite_dy(player) > MAX_SPEED)
  {
    sprite_set_dy(player, MAX_SPEED);
  }
  else if (sprite_dy(player) < -(MAX_SPEED))
  {
    sprite_set_dy(player, -(MAX_SPEED));
  }
}

void update_poles(game_data *game)
{
  int i;

  for (i = 0; i < NUM_POLES; i++)
  {
    update_sprite(game->poles[i].up_pole);
    update_sprite(game->poles[i].down_pole);

    if ((sprite_x(game->poles[i].up_pole) < sprite_x(game->player.sprite)) && (game->poles[i].score_limiter == true))
    {
      game->poles[i].score_limiter = false;
      game->player.score++;
    }

    if (sprite_offscreen(game->poles[i].up_pole) && sprite_offscreen(game->poles[i].down_pole) && (game->poles[i].score_limiter == false))
    {
      reset_pole_data(&game->poles[i]);
    }
  }
}

void update_background(background_data scene)
{
  update_sprite(scene.foreground);
  update_sprite(scene.foreroof);
  update_sprite(scene.background);

  if (sprite_x(scene.foreground) <= -(sprite_width(scene.foreground) / 2))
  {
    sprite_set_x(scene.foreground, 0);
    sprite_set_x(scene.foreroof, 0);
  }
  if (sprite_x(scene.background) <= -(sprite_width(scene.background) / 2))
  {
    sprite_set_x(scene.background, 0);
  }
}

void update_player(sprite player, player_state state)
{
  if (state == PLAY)
  {
    update_velocity(player);
  }
  update_sprite(player);
}

void update_game(game_data *game)
{
  if (!game->player.is_dead)
  {
    check_for_collisions(game);
    handle_input(game->player.sprite, &game->player.state);
    update_background(game->scene);
    update_player(game->player.sprite, game->player.state);
    if (game->player.state == PLAY)
    {
      update_poles(game);
    }
  }
  else
  {
    reset_game(game);
  }
}

void draw_poles(poles poles)
{
  int i;

  for (i = 0; i < NUM_POLES; i++)
  {
    draw_sprite(poles[i].up_pole);
    draw_sprite(poles[i].down_pole);
  }
}

void draw_game(game_data* game)
{
  char str[15];
  sprintf(str, "%d", game->player.score);

  draw_sprite(game->scene.background);
  draw_poles(game->poles);
  draw_sprite(game->scene.foreroof);
  draw_sprite(game->scene.foreground);
  draw_sprite(game->player.sprite);
  if (game->player.state == PLAY)
  {
    draw_text(str, COLOR_WHITE, "GameFont", 10, 0);
  }
  else if (game->player.state == MENU)
  {
    draw_bitmap(bitmap_named("Logo"), 0, 40);
    draw_text("PRESS SPACE!",
    COLOR_WHITE,
    "GameFont",
    screen_width() / 2 - text_width(font_named("GameFont"), "PRESS SOACE!") / 2,
    sprite_y(game->player.sprite) + text_height(font_named("GameFont"), " ") * 2);
  }
}

void set_up_game(game_data* game)
{
  int i;

  load_resource_bundle_named("CaveEscape", "CaveEscape.txt", false);
  for (i = 0; i < NUM_POLES; i++)
  {
    game->poles[i] = get_random_poles();
  }
  game->player = get_new_player();
  game->scene = get_new_background();
  fade_music_in("GameMusic", -1, 15000);
}

int main()
{
    game_data game;

    open_graphics_window("Cave Escape", 432, 768);
    set_up_game(&game);

    do
    {
      process_events();
      clear_screen(ColorWhite);
      update_game(&game);
      draw_game(&game);
      refresh_screen();

    } while(!window_close_requested());

    return 0;
}

File: Pascal/CaveEscape/src/CaveEscape10.pas
program GameMain;
uses SwinGame, sgTypes, sgTimers, sgSprites, sysUtils;

const
  GRAVITY = 0.08;
  MAX_SPEED = 5;
  JUMP_RECOVERY_BOOST = 2;
  FOREGROUND_FOREROOF_POLE_SCROLL_SPEED = -2;
  BACKGROUND_SCROLL_SPEED = -1;
  NUM_POLES = 4;

type
    PoleData = record
      ScoreLimiter: Boolean;
      UpPole: Sprite;
      DownPole: Sprite;
    end;

    Poles = array [0..NUM_POLES - 1] of PoleData;

    BackgroundData = record
      Foreroof: Sprite;
      Foreground: Sprite;
      Background: Sprite;
    end;

    PlayerState = (Menu, Play);

    Player = record
      Sprite: Sprite;
      Score: Integer;
      IsDead: Boolean;
      State: PlayerState;
    end;

    GameData = record
      Player: Player;
      Scene: BackgroundData;
      Poles: Poles;
    end;

function GetNewPlayer(): Player;
begin
  result.Sprite := CreateSprite(BitmapNamed('Player'), AnimationScriptNamed('PlayerAnimations'));
  SpriteSetX(result.Sprite, ScreenWidth() / 2 - SpriteWidth(result.Sprite));
  SpriteSetY(result.Sprite, ScreenHeight() / 2);
  SpriteStartAnimation(result.Sprite, 'Fly');
  result.Score := 0;
  result.IsDead := false;
  result.State := Menu;
end;

function GetRandomPoles(): PoleData;
begin
  result.UpPole := CreateSprite(BitmapNamed('UpPole'));
  result.DownPole := CreateSprite(BitmapNamed('DownPole'));
  SpriteSetX(result.UpPole, ScreenWidth() + RND(1200));
  SpriteSetY(result.UpPole, ScreenHeight() - SpriteHeight(result.UpPole) - RND(BitmapHeight(BitmapNamed('Foreground'))));
  SpriteSetX(result.DownPole, SpriteX(result.UpPole));
  SpriteSetY(result.DownPole, RND(BitmapHeight(BitmapNamed('Foreroof'))));;
  SpriteSetDx(result.UpPole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  SpriteSetDx(result.DownPole, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  result.ScoreLimiter := true;
end;

function GetNewBackground(): BackgroundData;
begin
  result.Background := CreateSprite(BitmapNamed('Background'));
  SpriteSetX(result.Background, 0);
  SpriteSetY(result.Background, 0);
  SpriteSetDx(result.Background, BACKGROUND_SCROLL_SPEED);

  result.Foreground := CreateSprite(BitmapNamed('Foreground'), AnimationScriptNamed('ForegroundAminations'));
  SpriteSetX(result.Foreground, 0);
  SpriteSetY(result.Foreground, ScreenHeight() - SpriteHeight(result.Foreground));
  SpriteSetDx(result.Foreground, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
  SpriteStartAnimation(result.Foreground, 'Fire');

  result.Foreroof := CreateSprite(BitmapNamed('Foreroof'));
  SpriteSetX(result.Foreroof, 0);
  SpriteSetY(result.Foreroof, 0);
  SpriteSetDx(result.Foreroof, FOREGROUND_FOREROOF_POLE_SCROLL_SPEED);
end;

procedure HandleInput(player: Player);
begin
  if KeyTyped(SpaceKey) and (player.State = Play) then
  begin
    SpriteSetDy(player.Sprite, SpriteDy(player.Sprite) - JUMP_RECOVERY_BOOST);
  end
  else if KeyTyped(SpaceKey) then
  begin
    player.State := Play;
  end;
end;

procedure CheckForCollisions(var game: GameData);
var
  i: Integer;
begin
  if (SpriteCollision(game.Player.Sprite, game.Scene.Foreground)) or (SpriteCollision(game.Player.Sprite, game.Scene.Foreroof)) then
  begin
    game.Player.IsDead := true;
    exit;
  end;

  for i := Low(game.Poles) to High(game.Poles) do
  begin
    if SpriteCollision(game.Player.Sprite, game.Poles[i].UpPole) or SpriteCollision(game.Player.Sprite, game.Poles[i].DownPole)then
    begin
      game.Player.IsDead := true;
      exit;
    end;
  end;
end;

procedure ResetPoleData(var poles: PoleData);
begin
  FreeSprite(poles.UpPole);
  FreeSprite(poles.DownPole);
  poles := GetRandomPoles();
end;

procedure ResetPlayer(var player: Player);
begin
  FreeSprite(player.Sprite);
  player := GetNewPlayer();
end;

procedure ResetGame(var game: GameData);
var
  i: Integer;
begin
  ResetPlayer(game.Player);
  for i:= Low(game.Poles) to High(game.Poles) do
  begin
    ResetPoleData(game.Poles[i]);
  end;
end;

procedure UpdateVelocity(player: Sprite);
begin
  SpriteSetDy(player, SpriteDy(player) + GRAVITY);

  if SpriteDy(player) > MAX_SPEED then
  begin
    SpriteSetDy(player, MAX_SPEED);
  end
  else if SpriteDy(player) < -(MAX_SPEED) then
  begin
    SpriteSetDy(player, -(MAX_SPEED));
  end;
end;

procedure UpdatePoles(var poles: PoleData; var player: Player);
begin
  UpdateSprite(poles.UpPole);
  UpdateSprite(poles.DownPole);

  if (SpriteX(poles.UpPole) < SpriteX(player.Sprite)) and (poles.ScoreLimiter = true) then
  begin
    if  then
    begin
      poles.ScoreLimiter := false;
      player.Score += 1;
    end;
  end;

  if ((SpriteX(poles.UpPole) + SpriteWidth(poles.UpPole)) < 0) and ((SpriteX(poles.DownPole) + SpriteWidth(poles.DownPole)) < 0) then
  begin
    ResetPoleData(poles);
  end;
end;

procedure UpdatePolesArray(var polesArray: Poles; var player: Player);
var
  i: Integer;
begin
  for i:= Low(polesArray) to High(polesArray) do
  begin
    UpdatePoles(polesArray[i], player);
  end;
end;

procedure UpdateBackground(scene: BackgroundData);
begin
  UpdateSprite(scene.ForeGround);
  UpdateSprite(scene.Foreroof);
  updateSprite(scene.Background);
  if (SpriteX(scene.Foreground) <= -(SpriteWidth(scene.ForeGround) / 2)) then
  begin
    SpriteSetX(scene.Foreground, 0);
    SpriteSetX(scene.Foreroof, 0);
  end;
  if (SpriteX(scene.Background) <= -(SpriteWidth(scene.Background) / 2)) then
  begin
    SpriteSetX(scene.Background, 0);
  end;
end;

procedure UpdatePlayer(player: Player);
begin
  if (player.State = Play) then
  begin
    UpdateVelocity(player.Sprite);
  end;
  UpdateSprite(player.Sprite);
end;

procedure UpdateGame(var game: GameData);
begin
  if not (game.Player.IsDead) then
  begin
    CheckForCollisions(game);
    HandleInput(game.Player);
    UpdateBackground(game.Scene);
    UpdatePlayer(game.Player);
    if (game.Player.State = Play) then
    begin
      UpdatePolesArray(game.Poles, game.Player);
    end;
  end
  else //The player has died :(
  begin
    ResetGame(game);
  end;
end;

procedure DrawPoles(poles: PoleData);
begin
  DrawSprite(poles.UpPole);
  DrawSprite(poles.DownPole);
end;

procedure DrawPolesArray(polesArray: Poles);
var
  i: Integer;
begin
  for i:= Low(polesArray) to High(polesArray) do
  begin
    DrawPoles(polesArray[i]);
  end;
end;

procedure DrawGame(const game: GameData);
begin
  DrawSprite(game.Scene.Background);
  DrawPolesArray(game.Poles);
  DrawSprite(game.Scene.Foreroof);
  DrawSprite(game.Scene.ForeGround);
  DrawSprite(game.Player.Sprite);
  if (game.Player.State = Play) then
  begin
    DrawText(IntToStr(game.Player.Score), ColorWhite, 'GameFont', 10, 0);
  end
  else if (game.Player.State = Menu) then
  begin
    DrawBitmap(BitmapNamed('Logo'), 0, 40);
    DrawText('PRESS SPACE!',
    ColorWhite,
    'GameFont',
    ScreenWidth() / 2 - TextWidth(FontNamed('GameFont'), 'PRESS SPACE!') / 2,
    SpriteY(game.Player.Sprite) + TextHeight(FontNamed('GameFont'), ' ') * 2);
  end;
end;

procedure SetUpGame(var game: GameData);
var
  i: Integer;
begin
  LoadResourceBundleNamed('CaveEscape', 'CaveEscape.txt', false);
  for i:= Low(game.Poles) to High(game.Poles) do
  begin
    game.Poles[i] := GetRandomPoles();
  end;
  game.Player := GetNewPlayer();
  game.Scene := GetNewBackground();
  FadeMusicIn('GameMusic', -1, 15000);
end;

procedure Main();
var
  game: GameData;
begin
  OpenGraphicsWindow('Cave Escape', 432, 768);
  SetUpGame(game);

  repeat // The game loop...
    ProcessEvents();
    ClearScreen(ColorWhite);
    UpdateGame(game);
    DrawGame(game);
    RefreshScreen();
  until WindowCloseRequested();
end;

begin
  Main();
end.
